### DISCUSSION TITLE
Adding favoriteStaticMembers for jdtls in LazyVim

### DISCUSSION THREAD

**Discussion by drone-ah** [LazyVim/LazyVim#3241] [Q&A]
I have been trying to add a couple of packages to this favorites list to make importing them easier, but witn no luck. Can someone help me figure out what i'm doing wrong?

I've tried various combinations, with the following being my final confid (which still doesn't work)

```
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        jdtls = {
          settings = {
            java = {
              completion = {
                favoriteStaticMembers = {
                  "org.assertj.core.api.Assertions.*",
                },
              },
            },
          },
        },
      },
    },
  },
  {
    "mfussenegger/nvim-jdtls",
    dependencies = { "folke/which-key.nvim" },
    ft = java_filetypes,
    opts = function()
      return {
        -- How to find the root dir for a given filename. The default comes from
        -- lspconfig which provides a function specifically for java projects.
        root_dir = require("lspconfig.server_configurations.jdtls").default_config.root_dir,

        -- How to find the project name for a given root dir.
        project_name = function(root_dir)
          return root_dir and vim.fs.basename(root_dir)
        end,

        -- Where are the config and workspace dirs for a project?
        jdtls_config_dir = function(project_name)
          return vim.fn.stdpath("cache") .. "/jdtls/" .. project_name .. "/config"
        end,
        jdtls_workspace_dir = function(project_name)
          return vim.fn.stdpath("cache") .. "/jdtls/" .. project_name .. "/workspace"
        end,

        -- How to run jdtls. This can be overridden to a full java command-line
        -- if the Python wrapper script doesn't suffice.
        cmd = { vim.fn.exepath("jdtls") },
        full_cmd = function(opts)
          local fname = vim.api.nvim_buf_get_name(0)
          local root_dir = opts.root_dir(fname)
          local project_name = opts.project_name(root_dir)
          local cmd = vim.deepcopy(opts.cmd)
          if project_name then
            vim.list_extend(cmd, {
              "-configuration",
              opts.jdtls_config_dir(project_name),
              "-data",
              opts.jdtls_workspace_dir(project_name),
            })
          end
          return cmd
        end,

        -- These depend on nvim-dap, but can additionally be disabled by setting false here.
        dap = { hotcodereplace = "auto", config_overrides = {} },
        dap_main = {},
        test = true,
        settings = {
          java = {
            completion = {
              favoriteStaticMembers = {
                "org.assertj.core.api.Assertions.*",
              },
            },
          },
        },
      }
    end,
  },
}

```

I am guessing that I need to configure one or the other, and I tried that first. It feels like the key for the configuration is wrong but the variations i've come up with to try have not worked.

Any help much appeciated.

Thank you

**Comment by dpetka2001** [comment]
I don't do Java at all, so please take anything I say with a grain of salt. Try the following
```lua
return {
  {
    "mfussenegger/nvim-jdtls",
    opts = {
      jdtls = function(opts)
        opts.settings = {
          java = {
           completion = {
              favoriteStaticMembers = {
                "org.assertj.core.api.Assertions.*",
              },
            },
          },
        }
        return opts
      end,
    },
  },
}
```
If it doesn't work out, unfortunately I don't know much more and you should wait for another user who does Java to help you.

**Reply by drone-ah** [reply]
Thank you for this. sadly though, it doesn't seem to work :(

**Reply by rogeriofrsouza** [reply]
Edit (June 5th 2024): 
It's possible to configure `jdtls` through `nvim-lspconfig`, there's a [default config for it](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#jdtls) 😝 
I thought you were using the Java Extra, the previous comment was specific for it. Sorry about that.

> I don't think it's possible to configure jdtls through nvim-lspconfig.
> As per the Java extra documentation indicates, it's only used to install the server via mason.
> https://www.lazyvim.org/extras/lang/java#nvim-lspconfig

**Reply by rogeriofrsouza** [reply]
@dpetka2001 answer is the correct way of configuring jdtls settings.
However, I don't know if this settings table gets merged with the one set in Java Extras:
https://github.com/LazyVim/LazyVim/blob/060f56d6d429c5ee3fd35f4fa0c152f4a42781ca/lua/lazyvim/plugins/extras/lang/java.lua#L121

**Reply by dpetka2001** [reply]
It should get merged as you can see [here](https://github.com/LazyVim/LazyVim/blob/060f56d6d429c5ee3fd35f4fa0c152f4a42781ca/lua/lazyvim/plugins/extras/lang/java.lua#L162-L171).

**Reply by rogeriofrsouza** [reply]
But when the function is executed here
https://github.com/LazyVim/LazyVim/blob/060f56d6d429c5ee3fd35f4fa0c152f4a42781ca/lua/lazyvim/plugins/extras/lang/java.lua#L9
A new settings table is assigned, right?
```lua
opts.settings = {
  java = {
    completion = {
      favoriteStaticMembers = {
        "org.assertj.core.api.Assertions.*",
      },
    },
  },
}
```

**Comment by drone-ah** [comment]
Further exploration shows that the behaviour is more nuanced than I expected. When writing code, it does autocomplete, and add in the import. 

However, when using code actions, it does not give me the option to import it. This lack of option to import is what I was testing and is why I thought it didn't work.

Thank you.

Any thoughts on how I might get it to work on code actions as well? I can work around it so it's less of an issue.

**Reply by rogeriofrsouza** [reply]
According to [this issue](https://github.com/mfussenegger/nvim-jdtls/issues/302), maybe that is the expected behavior?
`show up as completion candidates and to make organize imports resolve them`

I also did some experimenting here and code actions did show up for methods in `org.junit.jupiter.api.Assertions`, but not for `org.assertj.core.api.Assertions`.
![image](https://github.com/LazyVim/LazyVim/assets/77423511/529787be-2731-4331-ac4b-f22a7a9470fb)

Maybe it's a lack of support to Assertj from jdtls itself?
